controllers/platformoperator_controller.go

```diff
@@ -75,34 +79,71 @@ func (r *PlatformOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		}
 	}()
 
-	desiredBundle, err := r.Sourcer.Source(ctx, po)
+	bd, err := r.ensureDesiredBundleDeployment(ctx, po)
 	if err != nil {
+		// check whether we failed to return an active BundleDeployment
+		// resource due to sourcing failures. These sourcing failures are
+		// possible if the desired package name isn't present in the supported
+		// catalog sources in the cluster.
+		reason := platformtypes.ReasonInstallFailed
+		if errors.Is(err, errSourceFailed) {
+			reason = platformtypes.ReasonSourceFailed
+		}
 		meta.SetStatusCondition(&po.Status.Conditions, metav1.Condition{
 			Type:    platformtypes.TypeInstalled,
-			Status:  metav1.ConditionUnknown,
-			Reason:  platformtypes.ReasonSourceFailed,
+			Status:  metav1.ConditionFalse,
+			Reason:  reason,
 			Message: err.Error(),
 		})
 		return ctrl.Result{}, err
 	}
-	if err := r.Applier.Apply(ctx, po, desiredBundle); err != nil {
-		meta.SetStatusCondition(&po.Status.Conditions, metav1.Condition{
-			Type:    platformtypes.TypeInstalled,
-			Status:  metav1.ConditionUnknown,
-			Reason:  platformtypes.ReasonInstallFailed,
-			Message: err.Error(),
-		})
-		return ctrl.Result{}, err
+
+	// check whether the generated BundleDeployment are reporting any
+	// failures when attempting to unpack the configured registry+v1
+	// bundle contents, or persisting those unpack contents to the cluster.
+	if failureCond := util.InspectBundleDeployment(ctx, bd.Status.Conditions); failureCond != nil {
+		// avoid returning an error here as the controller is watching for BD resource
+		// events. this should avoid unnecessary requeues when the BD is still in the
+		// same state.
+		meta.SetStatusCondition(&po.Status.Conditions, *failureCond)
+		return ctrl.Result{}, nil
 	}
 	meta.SetStatusCondition(&po.Status.Conditions, metav1.Condition{
 		Type:    platformtypes.TypeInstalled,
 		Status:  metav1.ConditionTrue,
 		Reason:  platformtypes.ReasonInstallSuccessful,
-		Message: "Successfully applied the desired olm.bundle content",
+		Message: fmt.Sprintf("Successfully applied the %s BundleDeployment resource", bd.GetName()),
 	})
+	platformtypes.SetActiveBundleDeployment(po, bd.GetName())
+
 	return ctrl.Result{}, nil
 }
 
+func (r *PlatformOperatorReconciler) ensureDesiredBundleDeployment(ctx context.Context, po *platformv1alpha1.PlatformOperator) (*rukpakv1alpha1.BundleDeployment, error) {
+	bd := &rukpakv1alpha1.BundleDeployment{}
+
+	// check whether the underlying BD has already been generated to determine
+	// whether the sourcing logic needs to be run to avoid performing unnecessary
+	// work given upgrades aren't supported during phase 0. Note: this logic
+	// doesn't compare the current and desired status of the BD resource so it's
+	// possible that users/controllers/etc. can modify the generated BD resource.
+	// See https://github.com/openshift/platform-operators/issues/47 for more details.
+	if err := r.Get(ctx, types.NamespacedName{Name: po.GetName()}, bd); err != nil {
```

Comment:
Do we want to ensure that only one PO with a specific [ActiveBundleDeployment](https://github.com/openshift/api/blob/321abac6303f1c152f2de0cf853b5468ac403051/platform/v1alpha1/types.go#L35-L39) for a package in the catalog is present on cluster as well? Right now we're just checking for PO's that have the same name but those aren't necessarily indicative of their content.

Reply:
I just tested a situation where RukPak attempts to install two bundles that contain the same content and it looks like the underlying helm logic takes over and prevents that from going through. I think for phase 0 we're fine here in that case.

```
- lastTransitionTime: "2022-09-12T14:10:05Z"
    message: 'rendered manifests contain a resource that already exists. Unable to
      continue with install: Namespace "combo" in namespace "" exists and cannot be
      imported into the current release: invalid ownership metadata; annotation validation
      error: key "meta.helm.sh/release-name" must equal "combo-1": current value is
      "combo"'
```